Mathlib/Algebra/Lie/LieTheorem.lean

```diff
@@ -0,0 +1,558 @@
+/-
+Copyright (c) 2024 Lucas Whitfield. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Lucas Whitfield
+-/
+import Mathlib.Algebra.Lie.Solvable
+import Mathlib.Algebra.Lie.Weights.Linear
+
+/-!
+# Lie's theorem for Solvable Lie algebras.
+-/
+set_option autoImplicit false
+
+open LieAlgebra
+
+
+-- let k be a field of characteristic zero
+variable {k : Type*} [Field k] [CharZero k] [IsAlgClosed k]
+-- Let L be a Lie algebra over k
+variable {L : Type*} [LieRing L] [LieAlgebra k L]
+-- and let V be a finite-dimensional k-representation of L
+variable {V : Type*} [AddCommGroup V] [Module k V] [FiniteDimensional k V]
+  [LieRingModule L V] [LieModule k L V] [Nontrivial V]
+
+open FiniteDimensional
+
+/- We define the Submodules generated by repeatedly applying a linear map `f: V →ₗ[F] V`
+to a vector `v`-/
+section
+
+variable {F V : Type*} [Field F] [AddCommGroup V] [Module F V] [FiniteDimensional F V]
+variable (v : V) (f : Module.End F V)
+-- the Submodule generated by `v`, `f v`, `f f v`, ... , `f^[n-1] v`
+abbrev U (n : ℕ) : Submodule F V := Submodule.span F {v' : V | ∃ a < n, v' = (f^a) v}
+
+lemma U_zero_eq_bot : U v f 0 = ⊥ := by
+  rw [Submodule.span_eq_bot]
+  intro x ⟨a, ha, _⟩
+  contradiction
+
+lemma U_mono {a b : ℕ} (h : a ≤ b) : U v f a ≤ U v f b :=
+  Submodule.span_mono (fun _ ⟨c, hc, hw⟩ ↦ ⟨c, lt_of_lt_of_le hc h, hw⟩)
+
+lemma map_U_le (n : ℕ) : Submodule.map f (U v f n) ≤ U v f (n + 1) := by
+  rw [Submodule.map_span]
+  apply Submodule.span_mono
+  intro w ⟨z, ⟨m, hm, hz⟩, hw⟩
+  use m + 1, Nat.add_lt_add_right hm 1
+  rw [← hw, hz, ← LinearMap.mul_apply, ← pow_succ']
+
+
+
```

Comment:
The documentation linter will tell you this once this builds: please add a docstring to all new definitions.

Documentation for lemmas is optional, for theorems, you should also add one (but that's currently not enforced yet).